src/uct/tcp/tcp_ep.c

```diff
@@ -634,6 +682,13 @@ static unsigned uct_tcp_ep_progress_data_tx(uct_tcp_ep_t *ep)
         }
     }
 
+    if (ep->ctx_caps & UCS_BIT(UCT_TCP_EP_CTX_TYPE_PUT_RX_SENDING_ACK)) {
+        status = uct_tcp_ep_post_put_ack(ep);
+        if (status == UCS_OK) {
+            ep->ctx_caps &= ~UCS_BIT(UCT_TCP_EP_CTX_TYPE_PUT_RX_SENDING_ACK);
```

Comment:
looks like we can clear UCT_TCP_EP_CTX_TYPE_PUT_RX_SENDING_ACK bit unconditionally, before calling uct_tcp_ep_post_put_ack()

Reply:
no, we should clear this flag only if we can send data, otherwise, we should progress more until TX buffer is free
when `uct_tcp_ep_post_put_ack` returns UCS_OK, it means TX buffer was free and ACK was sent successfully

Reply:
but uct_tcp_ep_post_put_ack() will set the flag again if it fails to send..
it's weird because we don't clear the flag from uct_tcp_ep_put_rx_advance(). Maybe uct_tcp_ep_post_put_ack() should not return any status?